src/write/unit.rs

```diff
@@ -0,0 +1,1914 @@
+                self.format.word_size(),
+            )?;
+            w.write_u8(self.address_size)?;
+        } else if self.version == 5 {
+            w.write_u8(constants::DW_UT_compile.0)?;
+            w.write_u8(self.address_size)?;
+            w.write_offset(
+                abbrev_offset.0,
+                SectionId::DebugAbbrev,
+                self.format.word_size(),
+            )?;
+        } else {
+            return Err(Error::UnsupportedVersion(self.version));
+        }
+
+        self.entries[self.root.0].write(
+            w,
+            self,
+            &mut offsets,
+            abbrevs,
+            strings,
+            &mut unit_refs,
+            debug_info_refs,
+        )?;
+
+        let length = (w.len() - length_base) as u64;
+        w.write_initial_length_at(length_offset, length, self.format)?;
+
+        for (offset, entry) in unit_refs {
+            let entry_offset = offsets.entry(entry).0;
+            debug_assert_ne!(entry_offset, 0);
+            // This does not need relocation.
+            w.write_word_at(
+                offset.0,
+                (entry_offset - offsets.unit.0) as u64,
+                self.format.word_size(),
+            )?;
+        }
+
+        Ok(offsets)
+    }
+
+    /// Create an entry by reading the data in the given sections.
+    fn from<R: Reader<Offset = usize>>(
+        from_unit: &read::CompilationUnitHeader<R>,
+        units: &mut Vec<CompilationUnit>,
+        debug_abbrev: &read::DebugAbbrev<R>,
+        debug_str: &read::DebugStr<R>,
+        strings: &mut write::StringTable,
+        convert_address: &Fn(u64) -> Option<Address>,
+        unit_entry_offsets: &mut HashMap<DebugInfoOffset, (UnitId, UnitEntryId)>,
+    ) -> ConvertResult<UnitId> {
+        let id = UnitId(units.len());
+        let version = from_unit.version();
+        let address_size = from_unit.address_size();
+        let format = from_unit.format();
+        let mut entries = Vec::new();
+        let abbreviations = from_unit.abbreviations(debug_abbrev)?;
+        let mut from_tree = from_unit.entries_tree(&abbreviations, None)?;
+        let root = DebuggingInformationEntry::from(
+            from_tree.root()?,
+            from_unit,
+            &mut entries,
+            None,
+            id,
+            debug_str,
+            strings,
+            convert_address,
+            unit_entry_offsets,
+        )?;
+        units.push(CompilationUnit {
+            id,
+            version,
+            address_size,
+            format,
+            entries,
+            root,
+        });
+        Ok(id)
+    }
+}
+
+/// A Debugging Information Entry (DIE).
+///
+/// DIEs have a set of attributes and optionally have children DIEs as well.
+#[derive(Debug)]
+pub struct DebuggingInformationEntry {
+    id: UnitEntryId,
+    parent: Option<UnitEntryId>,
+    tag: constants::DwTag,
+    /// Whether to emit `DW_AT_sibling`.
+    // TODO: should this be automatic?
```

Comment:
What heuristics would we use to decide whether to emit this? Whether the children forest is of a certain size? It seems like maybe certain DIEs always end up having their children walked, while others are skipped past, so maybe there would be a whitelist of tags that should get the sibling links.

Anyways, agree that for now this should be manual.

Reply:
I was thinking maybe we could do it based on a tag whitelist. The whitelist members would be chosen based on typical access patterns of debuggers/unwinders.